File: jarvis/tools.py
"""Tool invocation with explicit permission gating.

Every tool declares a class: "safe" (read-only, always runs) or "sensitive"
(mutates local state — writes to the knowledge base, etc). Sensitive tools
only run if the caller's request explicitly opted in (`allow_write: true`).
Every call is appended to a local-only audit log — nothing here is ever
transmitted off-box, which is the privacy-protection half of this mechanism.
"""

import json
import re
import time

from jarvis.rag import _kb

AUDIT_LOG = _kb.root / "tools" / "audit.log"

TOOL_CALL_RE = re.compile(r"TOOL_CALL:\s*(\{)")

SYSTEM_PROMPT_TOOLS = """You have access to tools, but only these four —
never invent a tool name that isn't listed here:
- search_knowledge(query: str) — search the local knowledge base
- get_time() — current local date/time
- list_models() — list available local models
- add_note(title: str, content: str) — save a fact/note to the local knowledge base (requires write permission)

Only call a tool when the task genuinely needs one of these four things
(looking something up, saving a note, checking the time/model list). For
anything else — including math, reasoning, writing, or general knowledge —
answer directly yourself; do not call a tool.

To use one of the four tools, respond with ONLY a single line:
TOOL_CALL: {"name": "<tool_name>", "arguments": {...}}"""


def _audit(name, arguments, allowed, result_summary):
    AUDIT_LOG.parent.mkdir(parents=True, exist_ok=True)
    entry = {
        "ts": time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
        "tool": name,
        "arguments": arguments,
        "allowed": allowed,
        "result_summary": result_summary,
    }
    with AUDIT_LOG.open("a") as f:
        f.write(json.dumps(entry) + "\n")


def _t_search_knowledge(args):
    q = args.get("query", "")
    results = _kb.search(q, max_results=5)
    return {"results": results}


def _t_get_time(_args):
    return {"time": time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}


def _t_list_models(_args):
    from jarvis.routing import MODEL_ROUTING
    return {"models": list(MODEL_ROUTING.keys())}


def _t_add_note(args):
    title = args.get("title", "note")
    content = args.get("content", "")
    path = _kb.add_document(title, content)
    return {"saved": path}


TOOLS = {
    "search_knowledge": {"fn": _t_search_knowledge, "class": "safe"},
    "get_time": {"fn": _t_get_time, "class": "safe"},
    "list_models": {"fn": _t_list_models, "class": "safe"},
    "add_note": {"fn": _t_add_note, "class": "sensitive"},
}


def parse_tool_call(text):
    """Extract the first TOOL_CALL: {...} directive from model output, if
    present. Brace-matched rather than regex-captured to the closing brace —
    a model that emits multiple TOOL_CALL lines (small models do) would
    otherwise have its first call's JSON swallow everything up to the last
    line's closing brace, producing invalid JSON that silently fails to parse."""
    m = TOOL_CALL_RE.search(text or "")
    if not m:
        return None
    start = m.start(1)
    depth = 0
    for i, ch in enumerate(text[start:], start=start):
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                try:
                    call = json.loads(text[start:i + 1])
                except json.JSONDecodeError:
                    return None
                if "name" not in call:
                    return None
                call.setdefault("arguments", {})
                return call
    return None


def format_tool_followup(result, allowed):
    if not allowed:
        return (f"Tool call was denied ({result.get('error', 'not permitted')}). "
                f"Do not attempt another tool call — answer the original question "
                f"directly using your own knowledge instead.")
    return f"Tool result: {json.dumps(result)}\n\nNow answer the original question using this result."


def run_tool(name, arguments, allow_write=False):
    """Execute a tool under the permission gate. Always audited."""
    spec = TOOLS.get(name)
    if spec is None:
        _audit(name, arguments, False, "unknown tool")
        return {"error": f"unknown tool: {name}"}, False

    if spec["class"] == "sensitive" and not allow_write:
        _audit(name, arguments, False, "blocked: write permission not granted")
        return (
            {"error": f"tool '{name}' requires write permission (pass allow_write: true)"},
            False,
        )

    try:
        result = spec["fn"](arguments)
        _audit(name, arguments, True, str(result)[:200])
        return result, True
    except Exception as e:
        _audit(name, arguments, True, f"error: {e}")
        return {"error": str(e)}, True

File: tools/jarvis/tools.cpp
#include "tools.h"

#include <chrono>
#include <cstdio>
#include <ctime>
#include <fstream>
#include <map>

#include "routing.h"

using json = nlohmann::json;

namespace jarvis {

const char* SYSTEM_PROMPT_TOOLS =
    "You have access to exactly four tools. To use one, respond with ONLY a "
    "single line of the form:\n"
    "TOOL_CALL: {\"name\": \"<tool_name>\", \"arguments\": {...}}\n\n"
    "Available tools:\n"
    "- search_knowledge(query): search the local knowledge base.\n"
    "- get_time(): current UTC time.\n"
    "- list_models(): list available model ids.\n"
    "- add_note(title, content): save a note to the knowledge base (requires write permission).\n\n"
    "For anything else — math, reasoning, writing, general knowledge — answer directly, "
    "do not emit a TOOL_CALL.";

std::optional<ToolCall> parse_tool_call(const std::string& content) {
    static const std::string marker = "TOOL_CALL:";
    size_t marker_pos = content.find(marker);
    if (marker_pos == std::string::npos) return std::nullopt;

    size_t brace_start = content.find('{', marker_pos);
    if (brace_start == std::string::npos) return std::nullopt;

    // Brace-depth match to find the corresponding closing brace, tolerant of
    // braces inside string literals within the JSON.
    int depth = 0;
    bool in_string = false;
    bool escape = false;
    size_t i = brace_start;
    size_t brace_end = std::string::npos;
    for (; i < content.size(); i++) {
        char c = content[i];
        if (escape) { escape = false; continue; }
        if (c == '\\') { escape = true; continue; }
        if (c == '"') { in_string = !in_string; continue; }
        if (in_string) continue;
        if (c == '{') depth++;
        else if (c == '}') {
            depth--;
            if (depth == 0) { brace_end = i; break; }
        }
    }
    if (brace_end == std::string::npos) return std::nullopt;

    std::string blob = content.substr(brace_start, brace_end - brace_start + 1);
    json parsed;
    try {
        parsed = json::parse(blob);
    } catch (...) {
        return std::nullopt;
    }
    if (!parsed.contains("name") || !parsed["name"].is_string()) return std::nullopt;

    ToolCall call;
    call.name = parsed["name"].get<std::string>();
    call.arguments = parsed.value("arguments", json::object());
    return call;
}

namespace {

json tool_search_knowledge(KnowledgeBase& kb, const json& args) {
    std::string query = args.value("query", "");
    auto results = kb.search(query, 5);
    json arr = json::array();
    for (auto& r : results) arr.push_back({{"path", r.path}, {"title", r.title}, {"score", r.score}, {"snippet", r.snippet}});
    return {{"results", arr}};
}

json tool_get_time(KnowledgeBase&, const json&) {
    return {{"time", iso8601_now()}};
}

json tool_list_models(KnowledgeBase&, const json&) {
    return {{"models", model_ids()}};
}

json tool_add_note(KnowledgeBase& kb, const json& args) {
    std::string title = args.value("title", "note");
    std::string content = args.value("content", "");
    std::string path = kb.add_document(title, content);
    return {{"saved", path}};
}

struct ToolSpec {
    ToolClass cls;
    json (*fn)(KnowledgeBase&, const json&);
};

const std::map<std::string, ToolSpec>& tool_registry() {
    static const std::map<std::string, ToolSpec> reg = {
        {"search_knowledge", {ToolClass::Safe, tool_search_knowledge}},
        {"get_time", {ToolClass::Safe, tool_get_time}},
        {"list_models", {ToolClass::Safe, tool_list_models}},
        {"add_note", {ToolClass::Sensitive, tool_add_note}},
    };
    return reg;
}

void audit(KnowledgeBase& kb, const std::string& tool, const json& arguments, bool allowed,
           const std::string& result_summary) {
    std::filesystem::path log_path = kb.root() / "tools" / "audit.log";
    std::ofstream f(log_path, std::ios::binary | std::ios::app);
    if (!f) return;
    json entry = {
        {"ts", iso8601_now()},
        {"tool", tool},
        {"arguments", arguments},
        {"allowed", allowed},
        {"result_summary", result_summary.substr(0, 200)},
    };
    f << entry.dump() << "\n";
}

} // namespace

ToolRunResult run_tool(KnowledgeBase& kb, const std::string& name, const json& arguments, bool allow_write) {
    const auto& reg = tool_registry();
    auto it = reg.find(name);
    if (it == reg.end()) {
        json err = {{"error", "unknown tool: " + name}};
        audit(kb, name, arguments, false, err.dump());
        return {err, false};
    }

    if (it->second.cls == ToolClass::Sensitive && !allow_write) {
        json err = {{"error", "write permission denied for tool: " + name}};
        audit(kb, name, arguments, false, err.dump());
        return {err, false};
    }

    try {
        json result = it->second.fn(kb, arguments);
        audit(kb, name, arguments, true, result.dump());
        return {result, true};
    } catch (const std::exception& e) {
        json err = {{"error", std::string(e.what())}};
        audit(kb, name, arguments, true, err.dump());
        return {err, true};
    }
}

std::string format_tool_followup(const json& result, bool allowed) {
    if (!allowed) {
        std::string reason = result.value("error", "denied");
        return "The tool call was not permitted (" + reason +
               "). Do not retry it — answer from your own knowledge instead, "
               "and mention that the action requires write permission if relevant.";
    }
    return "The tool returned this result:\n" + result.dump() +
           "\nUse this result to answer the original question.";
}

} // namespace jarvis
